Add slice support (#233)
minirust/minirust: spec/lang/representation.md

Pointers are significantly more complex to represent than just the integer address.
For one, they need to encode the provenance.
When decoding, we have to deal with the possibility of the pointer bytes not all having the same provenance;
this is defined to yield a pointer without provenance.
On the other hand, some pointers are wide pointers which also need to encode their metadata.

> The helpers `decode_ptr` and `encode_ptr` deal with the thin pointers.

Reviewer: ```suggestion
The helpers `decode_ptr` and `encode_ptr` deal with thin pointers.
```

> For wide pointers `PtrType::as_wide_pair` defines the pointer as a pair of a thin pointer,

Reviewer: ```suggestion
For wide pointers, `PtrType::as_wide_pair` defines the pointer as a pair of a thin pointer and some metadata type.
```

> and some metadata type.

Reviewer: ```suggestion
```

this is define to yield a pointer without provenance.
On the other hand, some pointers are wide pointers which also need to encode their metadata.

> The helpers `decode_ptr` and `encode_ptr` deal with the thin pointers,

Reviewer: ```suggestion
The helpers `decode_ptr` and `encode_ptr` deal with the thin pointers.
```

> and the `decode` call them when the pointer is thin.

Reviewer: I can't quite parse this... should it be "calls them"? And no "the"?
But it also seems unnecessary to say this, the `decode` sources are right below. Saying they deal with thin pointers conveys all the relevant high-level parts, IMO.

    
}

impl PtrType {

>     // Returns a 2-tuple type representing this wide pointer or `None` if it is thin.

Reviewer: ```suggestion
    // Returns a pair type representing this wide pointer or `None` if it is thin.
```
Maybe call it `as_pair` then?

    fn as_tuple_ty<M: Memory>(self) -> Option<Type> {
        let meta_ty = match self.meta_kind() {

>             PointerMetaKind::None => throw!(),

Reviewer: We *could* put a unit type here and then treat thin and wide pointers more uniformly?

Author: Maybe, but keep in mind that there needs to be a separate "base case" for thin pointers anyway, because the Pointer::decode will be called by the Tuple::decode, meaning it is recursive. Preventing this by an explicit panic seemed like the least error prone solution to me.

Reviewer: Ah, good point. That should be noted in a comment somewhere.

But also, maybe there is a way here to structure this that does not require any unwraps -- something more like `if let Some(...)` then `if ...`?

>             PointerMetaKind::ElementCount => Type::Int(IntType { signed: Unsigned, size: M::T::PTR_SIZE }),

Reviewer: Seems reasonably to factor this into a `ty` method on `PointerMetaKind`.

        };
        let thin_pointer_field = (Offset::ZERO, Type::Ptr(PtrType::Raw { meta_kind: PointerMetaKind::None }));

>         let metadata_field = (Offset::from_bytes(M::T::PTR_SIZE.bytes()).expect("PTR_SIZE is positive"), meta_ty);

Reviewer: ```suggestion
        let metadata_field = (M::T::PTR_SIZE, meta_ty);
```

        ret(Type::Tuple {
            fields: list![thin_pointer_field, metadata_field],
            size: Int::from(2) * M::T::PTR_SIZE,

>             align: Align::from_bytes(M::T::PTR_SIZE.bytes()).expect("PTR_SIZE is aligned"),

Reviewer: ```suggestion
            // This anyway does not matter as we only use this type to encode/decode values.
            align: M::T::PTR_ALIGN,
```

        })
    }
}

impl PointerMeta {

>     fn from<M: Memory>(value: Value<M>) -> Option<PointerMeta> {

Reviewer: ```suggestion
    fn from_value<M: Memory>(value: Value<M>) -> Option<PointerMeta> {
```
For symmetry with `into_value`.

        match value {
            Value::Int(count) => Some(PointerMeta::ElementCount(count)),
            _ => None,
        }
    }

    fn into_value<M: Memory>(self) -> Value<M> {
        match self {
            PointerMeta::ElementCount(count) => Value::Int(count),
        }
    }
}

impl Type {
    fn decode<M: Memory>(Type::Ptr(ptr_type): Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
        if ptr_type.meta_kind() == PointerMetaKind::None {
            let ptr = decode_ptr::<M>(bytes)?;
            if !ptr_type.addr_valid(ptr.addr) {
                throw!();
            }
            ret(Value::Ptr(ptr.widen(None)))
        } else {
            let tuple_ty = ptr_type.as_tuple_ty::<M>().expect("ptr_type has metadata");
            let Value::Tuple(parts) = tuple_ty.decode::<M>(bytes)? else {
                panic!("as_tuple_ty always returns a tuple type");
            };
            let Value::Ptr(ptr) = parts[0] else {
                panic!("as_tuple_ty always returns tuple with the first field being a thin pointer");
            };

>             let meta = PointerMeta::from(parts[1]);

Reviewer: Shouldn't this panic if we get a `None`, since as_tuple_ty should always produce a suitable tuple type?

>             let meta = PointerMeta::from(parts[1]);

Author: ```suggestion
            let meta = PointerMeta::from_value(parts[1]);
```

}
```

> ### Unrepresentable types

Reviewer: ```suggestion
### Unsized types

Unsized types do not have values and thus there is no representation relation.
```
